catlearn/relation_cache.py

```diff
@@ -354,3 +356,58 @@ def match(
                 result_graph[src][tar].pop(NegativeMatch(best_match), None)
 
         return result_graph
+
+    def _prune_worst_relation(self) -> Optional[
+            CompositeArrow[NodeType, ArrowType]]:
+        """
+            Remove the relation with the lowest score in the cache.
+            Only remove 1st order arrows. arrows which are removed are those
+            with the lowest score relative to other arrows.
+
+            Returns the relation if it could be removed, None otherwise
+        """
+        # create a dictionary of total scores of each arrow
+        scores = {
+            arrow: torch.log(torch.sum(self[arrow]))
+            for arrow in self.arrows()}
+
+        # compute marginal utility of each arrow
+        utility = defaultdict(lambda: 0.)
+        for arrow in self.arrows():
+            src = arrow[0]
+            tar = arrow[-1]
+            other_scores = (
+                scores[arr] for arr in self.arrows(src, tar)
+                if arr is not arrow)
+
+            for idx in range(len(arrow)):
+                utility[arrow[idx:(idx + 1)]] += max(
+                    scores[arrow] - max(other_scores, default=-inf), 0.)
+
+        # identify worst relation
+        to_remove = min(utility, key=lambda arr: utility[arr], default=None)
+
+        # remove it from all dicts
+        if to_remove and isfinite(utility[to_remove]):
+            del self[to_remove]
+            return to_remove
```

Comment:
I'd rather always do `return to_remove` rather than relying on a missing return statement to do the right thing